### DISCUSSION TITLE
Efficient way get whole row of that corresponds to a min/max value of an expression

### DISCUSSION THREAD

**Discussion by thengineer** [duckdb/duckdb#6208] [Q&A] [answered]
I recently discovered `arg_min()` and `arg_max()`, which is close to what I'm looking for, but rather than returning a single field, I'd like to get the whole row returned. I tried `arg_min(arg, *)` and `arg_min(arg,  col1, col2)` but that doesnt work of course.

I could use the `ORDER BY arg ASC LIMIT 1` pattern, but that only works as long as I dont want to do any grouping.
In other DBMS I'd use the pattern with `ROW_NUMBER() OVER (ORDER BY arg ASC) as rn [..] WHERE rn=1` , but I feel like DuckDB has better (and faster!) alternatives.

Any ideas?

Thanks, Jake

**Comment by Mytherin** [comment, marked as answer]
You could run `arg_min` over a struct holding the entire row of the table:

```sql
create table tbl as select 42 i, 84 j;
select arg_min(tbl, i) from tbl;
┌──────────────────────────────┐
│     arg_min(integers, i)     │
│ struct(i integer, j integer) │
├──────────────────────────────┤
│ {'i': 42, 'j': 84}           │
└──────────────────────────────┘
```

I have not benchmarked it though - it may well not be faster than the alternatives you described.

**Reply by thengineer** [reply]
Thanks for the reply, is there a way to unpack the struct into columns? I only found UNNEST which seems to work on lists only.

**Reply by soerenwolfers** [reply]
you should be able to write `unnest(argmax(tbl, I))`.  maybe that functionality was only added in the almost two years since you've asked this

**Comment by Doekeb** [comment]
For anyone still looking for an answer, there is `arg_max(columns(*), val)` (and `arg_min`, and also `arg_max_null` and `arg_min_null`). For example:
```sql
D create table foo (a integer, b text);
D insert into foo values (1, 'a'), (2, 'b'), (3, null);
D select * from foo;
┌───────┬─────────┐
│   a   │    b    │
│ int32 │ varchar │
├───────┼─────────┤
│     1 │ a       │
│     2 │ b       │
│     3 │         │
└───────┴─────────┘
D select arg_max(columns(*), a) from foo;
┌───────┬─────────┐
│   a   │    b    │
│ int32 │ varchar │
├───────┼─────────┤
│     3 │ b       │
└───────┴─────────┘
D select arg_max_null(columns(*), a) from foo;
┌───────┬─────────┐
│   a   │    b    │
│ int32 │ varchar │
├───────┼─────────┤
│     3 │         │
└───────┴─────────┘
```

**Reply by soerenwolfers** [reply]
you should be able to write `unnest(argmax(tbl, i))` too
